HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+        params.update(kwds)
+
+        # Initialize a basic AgentType
+        IndShockConsumerType.__init__(self, verbose=verbose, quiet=quiet, **params)
+
+        self.solve_one_period = make_one_period_oo_solver(
+            ConsPerfForesightLabeledSolver
+        )
+
+    def update_solution_terminal(self):
+
+        u = UtilityFuncCRRA(self.CRRA)  # utility
+
+        mNrm = xr.DataArray(
+            np.append(0.0, self.aXtraGrid),
+            name="mNrm",
+            dims=("mNrm"),
+            attrs={"long_name": "cash_on_hand"},
+        )
+        state = xr.Dataset({"mNrm": mNrm})  # only one state var in this model
+
+        # optimal decision is to consume everything in the last period
+        cNrm = xr.DataArray(
+            mNrm,
+            name="cNrm",
+            dims=state.dims,
+            coords=state.coords,
+            attrs={"long_name": "consumption"},
+        )
+
+        v = u(cNrm)
+        v.name = "v"
+        v.attrs = {"long_name": "value function"}
+
+        v_der = u.der(cNrm)
+        v_der.name = "v_der"
+        v_der.attrs = {"long_name": "marginal value function"}
+
+        v_inv = cNrm.copy()
+        v_inv.name = "v_inv"
+        v_inv.attrs = {"long_name": "inverse value function"}
+
+        v_der_inv = cNrm.copy()
+        v_der_inv.name = "v_der_inv"
+        v_der_inv.attrs = {"long_name": "inverse marginal value function"}
+
+        dataset = xr.Dataset(
+            {
+                "cNrm": cNrm,
+                "v": v,
+                "v_der": v_der,
+                "v_inv": v_inv,
+                "v_der_inv": v_der_inv,
+            }
+        )
+
+        vfunc = ValueFuncCRRALabeled(
+            dataset[["v", "v_der", "v_inv", "v_der_inv"]], self.CRRA
+        )
+
+        self.solution_terminal = ConsumerSolutionLabeled(
+            value=vfunc,
+            policy=dataset[["cNrm"]],
+            continuation=None,
+            attrs={"m_nrm_min": 0.0},
+        )
+
+
+class ConsPerfForesightLabeledSolver(ConsIndShockSetup):
+    def def_utility_funcs(self):
+
+        self.u = UtilityFuncCRRA(self.CRRA)
+
+    def create_params_namespace(self):
+
+        self.params = SimpleNamespace(
+            Discount=self.DiscFac * self.LivPrb,
+            CRRA=self.CRRA,
+            Rfree=self.Rfree,
+            PermGroFac=self.PermGroFac,
+        )
+
+    def define_boundary_constraint(self):
+
+        # Calculate the minimum allowable value of money resources in this period
+        self.BoroCnstNat = (
+            self.solution_next.attrs["m_nrm_min"] - 1
+        ) / self.params.Rfree
+
+        # if natural borrowing constraint is binding constraint, then we can not
+        # evaluate the value function at that point, so we must fill out the data
+        if self.BoroCnstArt is None or self.BoroCnstArt <= self.BoroCnstNat:
+            self.m_nrm_min = self.BoroCnstNat
+            self.nat_boro_cnst = True
+
+            self.borocnst = xr.Dataset(
```

Comment:
I'm having trouble understanding why the borrowing constraint is represented as a DataSet.

Reply:
Just noting that you explained this to me and it made sense!